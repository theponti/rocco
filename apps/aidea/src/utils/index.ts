import { GetServerSidePropsContext } from "next";
import { getServerAuthSession } from "src/server/common/get-server-auth-session";

export async function getProtectedServerSideProps(
  ctx: GetServerSidePropsContext
) {
  const session = await getServerAuthSession(ctx);

  if (!session) {
    return {
      redirect: {
        destination: "/",
        // Use `false` to prevent browser caching
        permanent: false,
      },
    };
  }

  return {
    props: {
      session,
    },
  };
}
