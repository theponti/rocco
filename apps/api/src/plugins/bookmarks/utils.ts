import ogs from "open-graph-scraper";

interface GetOpenGraphDataParams {
  url: string;
}

interface GetSiteNameFallbackParams {
  url: string;
}

interface GetFaviconUrlFallbackParams {
  url: string;
}

interface OpenGraphData {
  description?: string;
  faviconUrl?: string;
  imageHeight?: number;
  imageWidth?: number;
  imageUrl?: string;
  siteName: string;
  title: string;
}

const twitterRegex =
  /(?:https?:\/\/)?(?:www\.)?\b(twitter\.com)\b((?:\/[a-z][a-z0-9_]*))?/i;
const applePodcastRegex =
  /(?:https?:\/\/)?\b(podcasts\.apple\.com)\b((?:\/[a-z][a-z0-9_]*))?/i;
const facebookRegex =
  /(?:https?:\/\/)?(?:www\.)?\b(m\.facebook|facebook|fb)\b\.(com|watch)\/?((?:\/[a-z][a-z0-9_]*))?/i;
const linkedinRegex =
  /(?:https?:\/\/)?(?:www\.)?\b(linkedin\.com)\b((?:\/[a-z][a-z0-9_]*))?/i;
// const spotifyRegex =
//   /https:\/\/open\.spotify\.com\/(track|album|playlist)\/([a-zA-Z0-9]+)/;
// const validBookmarkTypes = ["spotify", "airbnb"] as const;
// type ValidBookmarkType = (typeof validBookmarkTypes)[number];

function getFaviconUrlFallback({
  url,
}: GetFaviconUrlFallbackParams): string | undefined {
  switch (true) {
    case applePodcastRegex.test(url):
      return "https://podcasts.apple.com/favicon.ico";
    case twitterRegex.test(url):
      return "https://abs.twimg.com/favicons/twitter.2.ico";
    default:
      return undefined;
  }
}

// returns a fallback site name if the og:site_name is not available
function getSiteNameFallback({
  url,
}: GetSiteNameFallbackParams): string | undefined {
  switch (true) {
    case facebookRegex.test(url):
      return "Facebook";
    case linkedinRegex.test(url):
      return "LinkedIn";
    default:
      return undefined;
  }
}

function extractDomain(url: string): string {
  const domain = url.replace("http://", "").replace("https://", "");
  return domain.split(/[/?#]/)[0];
}

export async function getOpenGraphData({
  url,
}: GetOpenGraphDataParams): Promise<OpenGraphData> {
  const trimmedUrl = url.trim();
  const isTwitter = twitterRegex.test(trimmedUrl);

  const { error, result } = await ogs({
    onlyGetOpenGraphInfo: true,
    url: trimmedUrl,
  });

  if (error) {
    throw Error("Could not load data");
  }

  let ogImage = result.ogImage;
  let imageUrl: string | undefined;
  if (typeof ogImage === "object") {
    imageUrl = Array.isArray(ogImage) ? ogImage[0]?.url : "";
  } else if (typeof ogImage === "string") {
    imageUrl = ogImage;
  }

  let siteName: string | undefined = result.ogSiteName;
  if (!siteName) {
    siteName = getSiteNameFallback({ url: trimmedUrl });
  }

  // twitter returns '//abs.twimg.com/favicons/twitter.2.ico' for favicon
  // isTwitter handles it
  let faviconUrl: string | undefined = result.favicon;
  if (!faviconUrl || isTwitter) {
    faviconUrl = getFaviconUrlFallback({ url: trimmedUrl });
  }

  // if faviconUrl doesn't start with http or https or //, construct the faviconUrl
  // e.g. faviconUrl = '/favicon.ico' -> faviconUrl = 'https://www.example.com/favicon.ico'
  if (faviconUrl && !/^((https?:\/\/)|(\/\/))/.test(faviconUrl)) {
    faviconUrl = `https://${extractDomain(trimmedUrl)}/${
      faviconUrl.indexOf("/") === 0 ? faviconUrl.slice(1) : faviconUrl
    }`;
  }

  return {
    description: result.ogDescription,
    faviconUrl,
    imageUrl,
    siteName: siteName || extractDomain(trimmedUrl),
    title: result.ogTitle || extractDomain(trimmedUrl),
    imageWidth: result.ogImage ? result.ogImage[0]?.width : undefined,
    imageHeight: result.ogImage ? result.ogImage[0]?.height : undefined,
  };
}
