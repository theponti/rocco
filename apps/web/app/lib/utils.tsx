import { useAuth } from "@clerk/react-router";
import { LoadingScreen } from "app/components/Loading";
import { type ClassValue, clsx } from "clsx";
import { useEffect } from "react";
import { useNavigate } from "react-router";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
	return twMerge(clsx(inputs));
}

export function withAuth(Component: React.ComponentType) {
	return function AuthComponent(props: any) {
		const { isLoaded, userId } = useAuth();
		const navigate = useNavigate();

		useEffect(() => {
			if (!userId) {
				navigate("/login");
			}
		}, [userId, navigate]);

		if (isLoaded) {
			return <LoadingScreen />;
		}

		return <Component {...props} />;
	};
}
