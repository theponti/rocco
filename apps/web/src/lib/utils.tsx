import Loading from "@hominem/components/Loading";
import { type ClassValue, clsx } from "clsx";
import { useEffect } from "react";
import { type Navigate, useNavigate } from "react-router-dom";
import { twMerge } from "tailwind-merge";
import { useAuth } from "./auth";
import { AuthStatus } from "./auth/types";

export function cn(...inputs: ClassValue[]) {
	return twMerge(clsx(inputs));
}

export function withAuth(Component: React.ComponentType) {
	return function AuthComponent(props: any) {
		const { user, status } = useAuth();
		const navigate = useNavigate();

		useEffect(() => {
			if (!user) {
				navigate("/login");
			}
		}, [user, navigate]);

		if (status === AuthStatus.Unloaded) {
			return (
				<div className="flex w-full max-h-[300px] justify-center items-center">
					<Loading size="xl" />
				</div>
			);
		}

		return <Component {...props} />;
	};
}
