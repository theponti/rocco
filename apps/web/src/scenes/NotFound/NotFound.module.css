.wrap {
  display: flex;
  justify-content: center;
  background-color: rgba(106, 90, 205, 0.9);
  color: white;
  border-radius: 4px;
  padding: 24px;
  width: 100%;
  margin: 0 24px;
}

.wrap h1 {
  font-size: 48px;
  font-weight: 700;
  margin: 0;
}
