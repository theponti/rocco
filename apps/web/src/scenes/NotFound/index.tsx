import { Link, useLocation } from "react-router-dom";
import styles from "./NotFound.module.css";

function NotFound() {
  const pathname = useLocation().pathname;

  if (pathname.indexOf("invites") !== -1) {
    return (
      <div className={styles.wrap}>
        <h2 className="text-2xl font-semibold">
          This invite could not be found.
        </h2>
        <p className="text-md">
          If someone invited you to a list,{" "}
          <Link to="/login" className="text-blue-400 font-semibold">
            log in
          </Link>{" "}
          to view your new list.
        </p>
      </div>
    );
  }

  return (
    <div className={styles.wrap}>
      <h1>You lost, buddy?</h1>
    </div>
  );
}

export default NotFound;
