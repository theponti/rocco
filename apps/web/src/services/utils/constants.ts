export const CATEGORIES = [
  { name: "housing", percentage: 0.3 },
  { name: "savings", percentage: 0.25 },
  { name: "insurance", percentage: 0.1 },
  { name: "food", percentage: 0.1 },
  { name: "transportation", percentage: 0.1 },
  { name: "misc", percentage: 0.1 },
  { name: "utilities", percentage: 0.05 },
];

// eslint-disable-next-line
export default {
  CATEGORIES,
};
