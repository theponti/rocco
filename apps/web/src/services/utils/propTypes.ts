import PropTypes from "prop-types";

export type User = {
  id: string;
  email: string;
  name: string;
  isAdmin: string;
  createdAt: string;
  updatedAt: string;
};

export const userPropType = PropTypes.shape({
  id: PropTypes.string.isRequired,
  avatar: PropTypes.string,
  email: PropTypes.string.isRequired,
  name: PropTypes.string,
});

export const transactionPropType = PropTypes.shape({
  _id: PropTypes.string,
  payee: PropTypes.string,
  amount: PropTypes.number,
});
