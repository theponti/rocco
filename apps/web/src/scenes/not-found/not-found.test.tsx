import { screen } from "@testing-library/react";
import { useLocation, useMatch } from "react-router-dom";
import { renderWithProviders } from "src/test/utils";

import NotFound from ".";

describe("not-found", () => {
  it("should render not found for invites route", () => {
    (useLocation as jest.Mock).mockReturnValue({ pathname: "/invites/:id" });
    renderWithProviders(<NotFound />);

    expect(
      screen.getByText("This invite could not be found."),
    ).toMatchSnapshot();
  });

  it("should render not found for list route", () => {
    (useLocation as jest.Mock).mockReturnValue({ pathname: "/" });
    (useMatch as jest.Mock).mockImplementation((path) => {
      if (path === "/list/:id") {
        return true;
      }
      return false;
    });

    renderWithProviders(<NotFound />);

    expect(screen.getByText("This list could not be found.")).toMatchSnapshot();
  });
});
