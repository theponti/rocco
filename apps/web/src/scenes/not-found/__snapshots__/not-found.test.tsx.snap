// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`not-found > should render not found for invites route 1`] = `
<h2
  class="text-2xl font-semibold"
>
  This invite could not be found.
</h2>
`;

exports[`not-found > should render not found for list route 1`] = `
<h2
  class="text-2xl font-semibold"
>
  This list could not be found.
</h2>
`;
